javatests/arcs/android/storage/service/StorageServiceNgImplTest.kt

```diff
@@ -0,0 +1,245 @@
+    val deferredChannel1 = CompletableDeferred<IStorageChannel>()
+    val deferredChannel2 = CompletableDeferred<IStorageChannel>()
+
+    val storageServiceNg = createStorageService(this) as StorageServiceNgImpl
+    assertThat(storageServiceNg.stores.size).isEqualTo(0)
+
+    // Define the store options for two different stores.
+    val storeOptions1 = StoreOptions(RamDiskStorageKey("store1"), CountType())
+    val encodedStoreOptions1 = with(Parcel.obtain()) {
+      writeStoreOptions(storeOptions1, ParcelableCrdtType.Count, 0)
+      marshall()
+    }
+
+    val storeOptions2 = StoreOptions(RamDiskStorageKey("store2"), CountType())
+    val encodedStoreOptions2 = with(Parcel.obtain()) {
+      writeStoreOptions(storeOptions2, ParcelableCrdtType.Count, 0)
+      marshall()
+    }
+
+    // Open a storage channel for one store and assert that a new store has been created.
+    storageServiceNg.openStorageChannel(
+      encodedStoreOptions1,
+      createChannelCallback(deferredChannel1),
+      messageCallback
+    )
+    val channel1 = deferredChannel1.await()
+    assertThat(storageServiceNg.stores.size).isEqualTo(1)
+
+    // Open a storage channel for another store and assert that another new store has been created.
+    storageServiceNg.openStorageChannel(
+      encodedStoreOptions2,
+      createChannelCallback(deferredChannel2),
+      messageCallback
+    )
+    val channel2 = deferredChannel2.await()
+    assertThat(storageServiceNg.stores.size).isEqualTo(2)
+
+    // Also check that the two channels that were created are distinct.
+    assertThat(channel2).isNotSameInstanceAs(channel1)
+  }
+
+  @Test
+  fun openingTwoChannels_forSameStore_onlyCreatesOneStore() = runBlockingTest {
+    val deferredChannel1 = CompletableDeferred<IStorageChannel>()
+    val deferredChannel2 = CompletableDeferred<IStorageChannel>()
+
+    val storageServiceNg = createStorageService(this) as StorageServiceNgImpl
+    assertThat(storageServiceNg.stores.size).isEqualTo(0)
+
+    // Define the store options for a store.
+    val storeOptions1 = StoreOptions(RamDiskStorageKey("store1"), CountType())
+    val encodedStoreOptions1 = with(Parcel.obtain()) {
+      writeStoreOptions(storeOptions1, ParcelableCrdtType.Count, 0)
+      marshall()
+    }
+
+    // Open a storage channel for the store and assert that a new store has been created.
+    storageServiceNg.openStorageChannel(
+      encodedStoreOptions1,
+      createChannelCallback(deferredChannel1),
+      messageCallback
+    )
+    val channel1 = deferredChannel1.await()
+    assertThat(storageServiceNg.stores.size).isEqualTo(1)
+
+    // Open another storage channel for the same store. Ensure no new store has been created.
+    storageServiceNg.openStorageChannel(
+      encodedStoreOptions1,
+      createChannelCallback(deferredChannel2),
+      messageCallback
+    )
+    val channel2 = deferredChannel2.await()
+    assertThat(storageServiceNg.stores.size).isEqualTo(1)
```

Comment:
Keep this assertion, delete the one above

Reply:
With respect to the longer discussion above: is this really testing really verifying the behavior that we want?

For example, wouldn't it be possible to introduce a mistake in the channel later that would result in the second open creating a new store and replacing the existing one with the new one? This would result in each channel having a different store, but the test here would still succeed, right?

Reply:
Responded in longer discussion above ([here](https://github.com/PolymerLabs/arcs/pull/6603#discussion_r527391238))